File: protocols/proven-snmp/ffi/zig/src/snmp.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// snmp.zig — Zig FFI implementation of proven-snmp.
//
// Implements the SNMP session management primitive with:
//   - Slot-based session management (up to 64 concurrent sessions)
//   - Protocol version tracking (v1/v2c/v3)
//   - PDU type validation per version (v2c/v3-only PDU types)
//   - Error status management (RFC 3416)
//   - Variable binding count tracking
//   - PDU send counting
//   - Thread-safe via mutex on global state
//
// Tag values MUST match:
//   - Idris2 ABI (src/abi/Types.idr)
//   - C header   (generated/abi/snmp.h)

const std = @import("std");

// ── Enums (matching Idris2 SNMPABI.Types tag assignments exactly) ────────

/// Version — matches versionToTag
pub const Version = enum(u8) {
    v1 = 0,
    v2c = 1,
    v3 = 2,
};

/// PDUType — matches pduTypeToTag
pub const PDUType = enum(u8) {
    get_request = 0,
    get_next_request = 1,
    get_response = 2,
    set_request = 3,
    get_bulk_request = 4,
    inform_request = 5,
    snmpv2_trap = 6,
};

/// ErrorStatus — matches errorStatusToTag
pub const ErrorStatus = enum(u8) {
    no_error = 0,
    too_big = 1,
    no_such_name = 2,
    bad_value = 3,
    read_only = 4,
    gen_err = 5,
    no_access = 6,
    wrong_type = 7,
    wrong_length = 8,
    wrong_value = 9,
    no_creation = 10,
    inconsistent_value = 11,
    resource_unavailable = 12,
    commit_failed = 13,
    undo_failed = 14,
    authorization_error = 15,
};

/// SNMPError — error codes for FFI operations
pub const SNMPError = enum(u8) {
    ok = 0,
    invalid_slot = 1,
    not_active = 2,
    invalid_version = 3,
    invalid_pdu = 4,
    version_mismatch = 5,
    invalid_error_status = 6,
    varbind_limit = 7,
};

// ── Session Context ─────────────────────────────────────────────────────

/// Maximum variable bindings per PDU.
const MAX_VARBINDS: usize = 128;

const SessionCtx = struct {
    /// Whether this slot is in use.
    active: bool,
    /// SNMP protocol version.
    version: Version,
    /// Current error status.
    error_status: ErrorStatus,
    /// Number of PDUs sent in this session.
    pdu_count: u32,
    /// Current variable binding count.
    varbind_count: u32,
    /// Last PDU type sent (255 = none).
    last_pdu_type: u8,
};

// ── Global state (slot-based, mutex-protected) ──────────────────────────

const MAX_CONTEXTS: usize = 64;

const empty_ctx: SessionCtx = .{
    .active = false,
    .version = .v1,
    .error_status = .no_error,
    .pdu_count = 0,
    .varbind_count = 0,
    .last_pdu_type = 255,
};

var contexts: [MAX_CONTEXTS]SessionCtx = [_]SessionCtx{empty_ctx} ** MAX_CONTEXTS;
var mutex: std.Thread.Mutex = .{};

// ── Helpers ─────────────────────────────────────────────────────────────

/// Validate and return a pointer to an active context, or null.
fn getActive(slot: c_int) ?*SessionCtx {
    if (slot < 0 or slot >= MAX_CONTEXTS) return null;
    const idx: usize = @intCast(slot);
    if (!contexts[idx].active) return null;
    return &contexts[idx];
}

// ── ABI version ─────────────────────────────────────────────────────────

/// ABI version — must match SNMPABI.Foreign.abiVersion (currently 1).
pub export fn snmp_abi_version() callconv(.c) u32 {
    return 1;
}

// ── Lifecycle ───────────────────────────────────────────────────────────

/// Create a new SNMP session.
/// Returns slot index (0-63) or -1 if no slots available or invalid params.
pub export fn snmp_create(version: u8) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    // Validate version (0-2)
    if (version > 2) return -1;

    for (&contexts, 0..) |*ctx, i| {
        if (!ctx.active) {
            ctx.* = empty_ctx;
            ctx.active = true;
            ctx.version = @enumFromInt(version);
            return @intCast(i);
        }
    }
    return -1; // all slots occupied
}

/// Destroy an SNMP session, freeing its slot.
pub export fn snmp_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();

    if (slot < 0 or slot >= MAX_CONTEXTS) return;
    const idx: usize = @intCast(slot);
    contexts[idx].active = false;
}

// ── State queries ───────────────────────────────────────────────────────

/// Get the SNMP version tag for a slot.
/// Returns V1 (0) for invalid/inactive slots.
pub export fn snmp_get_version(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return @intFromEnum(ctx.version);
}

/// Get the current error status tag for a slot.
/// Returns NoError (0) for invalid/inactive slots.
pub export fn snmp_get_error(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return @intFromEnum(ctx.error_status);
}

/// Get the number of PDUs sent in this session.
pub export fn snmp_get_pdu_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return ctx.pdu_count;
}

/// Get the current variable binding count.
pub export fn snmp_get_varbind_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 0;
    return ctx.varbind_count;
}

/// Get the last PDU type sent (255 = none).
pub export fn snmp_get_last_pdu_type(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return 255;
    return ctx.last_pdu_type;
}

// ── Operations ──────────────────────────────────────────────────────────

/// Set the error status for a session.
/// Returns SNMPError tag.
pub export fn snmp_set_error(slot: c_int, err: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(SNMPError.invalid_slot);

    if (err > 15) {
        return @intFromEnum(SNMPError.invalid_error_status);
    }

    ctx.error_status = @enumFromInt(err);
    return @intFromEnum(SNMPError.ok);
}

/// Send a PDU of the given type.
/// Validates that the PDU type is valid for the session's SNMP version.
/// Returns SNMPError tag.
pub export fn snmp_send_pdu(slot: c_int, pdu: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(SNMPError.invalid_slot);

    // Validate PDU type (0-6)
    if (pdu > 6) {
        return @intFromEnum(SNMPError.invalid_pdu);
    }

    // Check version compatibility
    if (snmp_can_send_pdu(@intFromEnum(ctx.version), pdu) == 0) {
        return @intFromEnum(SNMPError.version_mismatch);
    }

    ctx.last_pdu_type = pdu;
    ctx.pdu_count += 1;
    return @intFromEnum(SNMPError.ok);
}

/// Add a variable binding to the current PDU.
/// Returns SNMPError tag.
pub export fn snmp_add_varbind(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(SNMPError.invalid_slot);

    if (ctx.varbind_count >= MAX_VARBINDS) {
        return @intFromEnum(SNMPError.varbind_limit);
    }

    ctx.varbind_count += 1;
    return @intFromEnum(SNMPError.ok);
}

/// Clear all variable bindings.
pub export fn snmp_clear_varbinds(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return;
    ctx.varbind_count = 0;
}

/// Set the SNMP version for a session.
/// Returns SNMPError tag.
pub export fn snmp_set_version(slot: c_int, ver: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const ctx = getActive(slot) orelse return @intFromEnum(SNMPError.invalid_slot);

    if (ver > 2) {
        return @intFromEnum(SNMPError.invalid_version);
    }

    ctx.version = @enumFromInt(ver);
    return @intFromEnum(SNMPError.ok);
}

// ── Stateless PDU-version validation ────────────────────────────────────

/// Check whether a PDU type is valid for a given SNMP version.
/// Returns 1 if valid, 0 if not.
/// V1 (0): GetRequest(0), GetNextRequest(1), GetResponse(2), SetRequest(3)
/// V2c (1) and V3 (2): all 7 PDU types
pub export fn snmp_can_send_pdu(version: u8, pdu: u8) callconv(.c) u8 {
    // All versions support PDU types 0-3
    if (pdu <= 3) return 1;
    // PDU types 4-6 require v2c (1) or v3 (2)
    if (pdu <= 6 and version >= 1) return 1;
    return 0;
}

File: bindings/elixir/lib/proven_servers/snmp.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
defmodule ProvenServers.Snmp do
  @moduledoc """
  SNMP protocol types for the proven-servers ABI.
  
  Mirrors the Idris2 module `SNMPABI.Types` and its type definitions:
  - `Version`     — SNMP protocol versions (3 constructors, tags 0-2)
  - `PduType`     — SNMP PDU types (7 constructors, tags 0-6)
  - `ErrorStatus` — SNMP error status codes (16 constructors, tags 0-15)
  
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  @doc "Standard SNMP agent port (RFC 3411)."
  @spec snmp_port() :: non_neg_integer()
  def snmp_port, do: 161

  @doc "Standard SNMP trap port (RFC 3411)."
  @spec snmp_trap_port() :: non_neg_integer()
  def snmp_trap_port, do: 162

  # ===========================================================================
  # Version (tags 0-2)
  # ===========================================================================

  @typedoc """
  Version types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type version :: :v1 | :v2c | :v3

  @version_tags %{
    v1: 0,
    v2c: 1,
    v3: 2,
  }

  @tag_to_version Map.new(@version_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `Version` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..2, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Snmp.version_from_tag(0)
      {:ok, :v1}
  """
  @spec version_from_tag(non_neg_integer()) :: {:ok, version()} | :error
  def version_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 2 do
    {:ok, Map.fetch!(@tag_to_version, tag)}
  end

  def version_from_tag(_tag), do: :error

  @doc """
  Encode a `Version` to the C-ABI tag value.
  """
  @spec version_to_tag(version()) :: non_neg_integer()
  def version_to_tag(val) when is_map_key(@version_tags, val) do
    Map.fetch!(@version_tags, val)
  end

  @doc """
  All `Version` variants in tag order.
  """
  @spec all_versions() :: [version()]
  def all_versions, do: [:v1, :v2c, :v3]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this version supports the User-based Security Model (USM).
  """
  @spec has_usm?(version()) :: boolean()
  def has_usm?(val) when val in [:v3], do: true
  def has_usm?(_val), do: false

  @doc """
  Whether this version uses community strings for authentication.
  """
  @spec uses_community_strings?(version()) :: boolean()
  def uses_community_strings?(val) when val in [:v1, :v2c], do: true
  def uses_community_strings?(_val), do: false

  @doc """
  Whether this version supports GetBulkRequest.
  """
  @spec supports_get_bulk?(version()) :: boolean()
  def supports_get_bulk?(val) when val in [:v1], do: false
  def supports_get_bulk?(_val), do: true

  # ===========================================================================
  # PduType (tags 0-6)
  # ===========================================================================

  @typedoc """
  PduType types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type pdu_type ::
          :get_request
          | :get_next_request
          | :get_response
          | :set_request
          | :get_bulk_request
          | :inform_request
          | :snmp_v2_trap

  @pdu_type_tags %{
    get_request: 0,
    get_next_request: 1,
    get_response: 2,
    set_request: 3,
    get_bulk_request: 4,
    inform_request: 5,
    snmp_v2_trap: 6,
  }

  @tag_to_pdu_type Map.new(@pdu_type_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `PduType` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..6, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Snmp.pdu_type_from_tag(0)
      {:ok, :get_request}
  """
  @spec pdu_type_from_tag(non_neg_integer()) :: {:ok, pdu_type()} | :error
  def pdu_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 6 do
    {:ok, Map.fetch!(@tag_to_pdu_type, tag)}
  end

  def pdu_type_from_tag(_tag), do: :error

  @doc """
  Encode a `PduType` to the C-ABI tag value.
  """
  @spec pdu_type_to_tag(pdu_type()) :: non_neg_integer()
  def pdu_type_to_tag(val) when is_map_key(@pdu_type_tags, val) do
    Map.fetch!(@pdu_type_tags, val)
  end

  @doc """
  All `PduType` variants in tag order.
  """
  @spec all_pdu_types() :: [pdu_type()]
  def all_pdu_types do
    [
      :get_request, :get_next_request, :get_response, :set_request, :get_bulk_request,
      :inform_request, :snmp_v2_trap
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this PDU is a request from manager to agent.
  """
  @spec is_request?(pdu_type()) :: boolean()
  def is_request?(val) when val in [:get_request, :get_next_request, :set_request, :get_bulk_request], do: true
  def is_request?(_val), do: false

  @doc """
  Whether this PDU is a notification (trap or inform).
  """
  @spec is_notification?(pdu_type()) :: boolean()
  def is_notification?(val) when val in [:inform_request, :snmp_v2_trap], do: true
  def is_notification?(_val), do: false

  @doc """
  Whether this PDU modifies agent state.
  """
  @spec is_write?(pdu_type()) :: boolean()
  def is_write?(val) when val in [:set_request], do: true
  def is_write?(_val), do: false

  # ===========================================================================
  # ErrorStatus (tags 0-15)
  # ===========================================================================

  @typedoc """
  ErrorStatus types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type error_status ::
          :no_error
          | :too_big
          | :no_such_name
          | :bad_value
          | :read_only
          | :gen_err
          | :no_access
          | :wrong_type
          | :wrong_length
          | :wrong_value
          | :no_creation
          | :inconsistent_value
          | :resource_unavailable
          | :commit_failed
          | :undo_failed
          | :authorization_error

  @error_status_tags %{
    no_error: 0,
    too_big: 1,
    no_such_name: 2,
    bad_value: 3,
    read_only: 4,
    gen_err: 5,
    no_access: 6,
    wrong_type: 7,
    wrong_length: 8,
    wrong_value: 9,
    no_creation: 10,
    inconsistent_value: 11,
    resource_unavailable: 12,
    commit_failed: 13,
    undo_failed: 14,
    authorization_error: 15,
  }

  @tag_to_error_status Map.new(@error_status_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ErrorStatus` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..15, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Snmp.error_status_from_tag(0)
      {:ok, :no_error}
  """
  @spec error_status_from_tag(non_neg_integer()) :: {:ok, error_status()} | :error
  def error_status_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 15 do
    {:ok, Map.fetch!(@tag_to_error_status, tag)}
  end

  def error_status_from_tag(_tag), do: :error

  @doc """
  Encode a `ErrorStatus` to the C-ABI tag value.
  """
  @spec error_status_to_tag(error_status()) :: non_neg_integer()
  def error_status_to_tag(val) when is_map_key(@error_status_tags, val) do
    Map.fetch!(@error_status_tags, val)
  end

  @doc """
  All `ErrorStatus` variants in tag order.
  """
  @spec all_error_statuss() :: [error_status()]
  def all_error_statuss do
    [
      :no_error, :too_big, :no_such_name, :bad_value, :read_only, :gen_err,
      :no_access, :wrong_type, :wrong_length, :wrong_value, :no_creation,
      :inconsistent_value, :resource_unavailable, :commit_failed, :undo_failed,
      :authorization_error
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this status indicates success.
  """
  @spec is_success?(error_status()) :: boolean()
  def is_success?(val) when val in [:no_error], do: true
  def is_success?(_val), do: false

  @doc """
  Whether this is an SNMPv1-only error code.
  """
  @spec is_v1_only?(error_status()) :: boolean()
  def is_v1_only?(val) when val in [:no_such_name, :bad_value, :read_only], do: true
  def is_v1_only?(_val), do: false

  @doc """
  Whether this error relates to authorisation/access control.
  """
  @spec is_auth_error?(error_status()) :: boolean()
  def is_auth_error?(val) when val in [:no_access, :authorization_error], do: true
  def is_auth_error?(_val), do: false

end
